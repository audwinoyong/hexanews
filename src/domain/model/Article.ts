/**
 * Model for the article.
 */
export type Article = {
  title: string,
  description: string
  category?: string
  authorId?: string
}