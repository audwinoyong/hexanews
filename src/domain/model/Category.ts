/**
 * Model for category object in our application.
 */

export type Category = {
  _id?: string,
  title: string,
  description: string,
}