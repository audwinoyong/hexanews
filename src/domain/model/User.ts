/**
 * Model for the user.
 */
export type User = {
  _id: string,
  email: string,
  name: string,
  description: string,
  createdAt: string,
}