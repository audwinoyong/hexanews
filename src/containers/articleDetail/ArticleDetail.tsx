/**
 * Display article detail container containing the article description
 */

import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { map, find, head, startCase } from 'lodash'
import { push } from 'connected-react-router'
import { translate } from 'react-i18next'
import * as moment from 'moment'
import * as React from 'react'
import Avatar from '@material-ui/core/Avatar'
import Button from '@material-ui/core/Button'
import CircularProgress from '@material-ui/core/CircularProgress'
import Divider from '@material-ui/core/Divider'
import Grid from '@material-ui/core/Grid'
import ReactHtmlParser from 'react-html-parser'
import Typography from '@material-ui/core/Typography'

import { DATE_FORMAT, HEADER_LINK } from '../../constants/config'
import { NotFound } from '../notFound/NotFound'
import { User } from '../../domain/model/User'
import * as actions from './actions'
import FooterCard from '../../components/footerCard/FooterCard'
import i18n from '../../i18n'
import ScrollRestoration from './ScrollRestoration'
import selector, { StateProps } from './selector'
import styles from './styles'
import SuccessSnackbar from '../../components/successSnackbar/SuccessSnackbar'

/**
 * All props required by the container
 */
type Props = {
  dispatch: Dispatch<any>
} & StateProps

/**
 * All state required by the container
 */
interface ComponentState {
  isDeleteSnackbarOpen: boolean
}

export class ArticleDetail extends React.Component<Props, ComponentState> {

  /**
   * Main constructor for Article Detail container and initialize state
   * @param props props required in the components
   */
  constructor(props: Props) {
    super(props)
    this.state = {
      isDeleteSnackbarOpen: false,
    }
  }

  /**
   * Function to close snackbar after delete article success
   */
  handleDeleteClose = () => {
    this.setState({ isDeleteSnackbarOpen: false })
  }

  /**
   * Function that render the article description
   */
  renderArticle = () => {
    const { dispatch, users, userArticle, isUserArticle, isDeletingArticle } = this.props
    if (!!userArticle) {
      const authorId = userArticle.authorId
      const author = find(users, (user: User) => user._id === authorId)
      const authorName = author ? author.name : i18n.t('articleDetail.authorDeleted')
      const initials = head(startCase(authorName))
      return (
        <Grid style={styles.contentContainer}>
          <Grid style={styles.profileContainer}>
            <Avatar style={styles.avatar}>
              {initials}
            </Avatar>
            <Grid style={styles.detailContainer as any}>
              <Typography
                variant='subheading'
                style={styles.profileName}
              >
                {authorName}
              </Typography>
              <Typography
                variant='body1'
                color='textSecondary'
              >
                {moment(userArticle.createdAt).format(DATE_FORMAT)}
              </Typography>
            </Grid>
            <Grid style={styles.buttonContainer}>
              {isUserArticle &&
                <Button
                  variant='outlined'
                  size='small'
                  component='button'
                  style={styles.editArticle}
                  onClick={() => dispatch(push(`../editArticle/${userArticle._id}`))}
                >
                  {i18n.t('articleDetail.editArticle')}
                </Button>
              }
              {isUserArticle &&
                <Button
                  variant='outlined'
                  size='small'
                  component='button'
                  style={styles.deleteArticle}
                  onClick={this.onDeleteArticle}
                >
                  {isDeletingArticle ? <CircularProgress size={22} /> : i18n.t('articleDetail.deleteArticle')}
                </Button>
              }
            </Grid>
          </Grid>
          <Typography
            variant='display1'
            component='h1'
            gutterBottom
          >
            {userArticle.title}
          </Typography>
          {ReactHtmlParser(userArticle.description)}
        </Grid>
      )
    }
    return
  }

  /**
   * Function that render footer cards for related articles
   */
  renderFooterCards = () => {
    const { dispatch, footerArticles, users } = this.props
    return map(footerArticles, (article) => {
      const author = find(users, (user: User) => user._id === article.authorId)
      const authorName = author ? author.name : i18n.t('articleDetail.authorDeleted')
      return (
        <Grid item xs={12} md={4} key={article._id}>
          <FooterCard
            article={article}
            authorName={authorName}
            dispatch={dispatch}
            key={article._id}
          />
        </Grid>
      )
    })
  }

  /**
   * Function that call deleteArticle action to delete the article
   */
  onDeleteArticle = async () => {
    const { dispatch, userArticle } = this.props
    const articleId = userArticle._id
    if (!!articleId) {
      this.setState({ isDeleteSnackbarOpen: true })
      setTimeout(async () => {
        await dispatch(actions.deleteArticle(articleId))
        dispatch(push(HEADER_LINK.profile))
      }, 500)
    }
  }

  /**
   * Render Article Detail container
   */
  public render() {
    const { dispatch, userArticle } = this.props
    const { isDeleteSnackbarOpen } = this.state

    if (!userArticle) {
      return (
        <NotFound
          dispatch={dispatch}
        />
      )
    }

    return (
      <ScrollRestoration>
        <Grid container style={styles.container}>
          <Grid item md={2} xs={1} />
          <Grid item md={8} xs={10}>
            <Grid container>
              {this.renderArticle()}
            </Grid>
            <Divider style={styles.footerDivider} />
            <Grid container spacing={24}>
              {this.renderFooterCards()}
            </Grid>
          </Grid>
          <Grid item md={2} xs={1} />
          <SuccessSnackbar
            isSnackbarOpen={isDeleteSnackbarOpen}
            message={i18n.t('articleDetail.deleteArticleSuccess')}
            handleClose={this.handleDeleteClose}
          />
        </Grid>
      </ScrollRestoration>
    )
  }
}

const ConnectedArticleDetail = connect(selector)(ArticleDetail)

export default translate('translations')(ConnectedArticleDetail)