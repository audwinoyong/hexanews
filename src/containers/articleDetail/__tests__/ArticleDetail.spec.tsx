import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'
import { omit } from 'lodash'

import { ArticleDetail } from '../ArticleDetail'
import { Article } from '../../../domain/model/Article'

Enzyme.configure({ adapter: new Adapter() })

const mockArticle: Article = {
  _id: '2',
  title: 'def',
  description: 'tester',
  category: 'tester',
  authorId: 'tester',
}

const mockUser = {
  _id: '1',
  createdAt: 'created',
  description: 'desc',
  email: 'user@email.com',
  name: 'user'
}

describe('Article Detail', () => {
  it('renders Article Detail', () => {
    const component = renderer.create(
      <ArticleDetail
        articles={[mockArticle]}
        users={[mockUser]}
        userArticle={mockArticle}
        isDeletingArticle={false}
        isUserArticle={true}
        article={mockArticle}
        footerArticles={[mockArticle]}
        dispatch={jest.fn()}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Article Detail with isDeletingArticle to be true', () => {
    const component = renderer.create(
      <ArticleDetail
        articles={[mockArticle]}
        users={[mockUser]}
        userArticle={mockArticle}
        isDeletingArticle={true}
        isUserArticle={true}
        footerArticles={[mockArticle]}
        article={mockArticle}
        dispatch={jest.fn()}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle onDeleteArticle action correctly', () => {
    const articleWithoutId = omit(mockArticle, ['_id'])
    const dispatch = jest.fn()
    const component = (
      <ArticleDetail
        articles={[mockArticle]}
        users={[mockUser]}
        userArticle={articleWithoutId}
        isDeletingArticle={false}
        isUserArticle={true}
        footerArticles={[mockArticle]}
        article={articleWithoutId}
        dispatch={dispatch}
      />
    )

    const layout = shallow(component)
    const instance: ArticleDetail = layout.instance() as ArticleDetail

    instance.onDeleteArticle()
    layout.update()

    expect(dispatch).not.toHaveBeenCalled()

    layout.setProps({ userArticle: mockArticle })
    layout.update()

    instance.onDeleteArticle()
    layout.update()

    expect(dispatch).toHaveBeenCalled()
  })
})