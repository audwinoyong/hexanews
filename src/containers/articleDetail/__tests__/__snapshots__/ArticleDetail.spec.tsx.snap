// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Detail renders Article Detail 1`] = `
<div
  className="MuiGrid-container-1"
  style={
    Object {
      "marginTop": "4%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-2-59"
  />
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-8-65"
  >
    <div
      className="MuiGrid-container-1"
    >
      <div
        style={
          Object {
            "marginBottom": "2%",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "marginBottom": "3%",
            }
          }
        >
          <div
            className="MuiAvatar-root-98 MuiAvatar-colorDefault-99"
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "green",
                "height": 50,
                "margin": 10,
                "width": 50,
              }
            }
          >
            A
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
              }
            }
          >
            <h3
              className="MuiTypography-root-101 MuiTypography-subheading-108"
              style={
                Object {
                  "marginTop": 12,
                }
              }
            >
              Author is deleted
            </h3>
            <p
              className="MuiTypography-root-101 MuiTypography-body1-110 MuiTypography-colorTextSecondary-124"
            >
              22 September 2018
            </p>
          </div>
          <div
            style={
              Object {
                "alignItems": "flex-end",
                "display": "flex",
                "flex": 1,
                "justifyContent": "flex-end",
              }
            }
          >
            <button
              className="MuiButtonBase-root-152 MuiButton-root-126 MuiButton-text-128 MuiButton-outlined-134 MuiButton-sizeSmall-149"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                  "marginRight": 10,
                }
              }
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-127"
              >
                Edit article
              </span>
              <span
                className="MuiTouchRipple-root-155"
              />
            </button>
            <button
              className="MuiButtonBase-root-152 MuiButton-root-126 MuiButton-text-128 MuiButton-outlined-134 MuiButton-sizeSmall-149"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                }
              }
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-127"
              >
                Delete article
              </span>
              <span
                className="MuiTouchRipple-root-155"
              />
            </button>
          </div>
        </div>
        <h1
          className="MuiTypography-root-101 MuiTypography-display1-105 MuiTypography-gutterBottom-118"
        >
          def
        </h1>
        tester
      </div>
    </div>
    <hr
      className="MuiDivider-root-162"
      style={
        Object {
          "marginBottom": 32,
          "marginTop": 32,
        }
      }
    />
    <div
      className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 MuiGrid-grid-md-4-61"
      >
        <div
          id="footer-2"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
              "flexGrow": 1,
            }
          }
        >
          <div
            className="MuiPaper-root-167 MuiPaper-elevation1-170 MuiPaper-rounded-168 MuiCard-root-166"
            style={
              Object {
                "boxShadow": "0px 0.25px 0.35px 0.35px rgba(0, 0, 0, 0.1)",
                "maxWidth": 385,
              }
            }
          >
            <div
              className="MuiCardMedia-root-194"
              src={undefined}
              style={
                Object {
                  "height": 0,
                  "paddingTop": "56.25%",
                }
              }
            />
            <div
              className="MuiCardContent-root-196"
              style={
                Object {
                  "paddingRight": "2%",
                }
              }
            >
              <h2
                className="MuiTypography-root-101 MuiTypography-caption-111 MuiTypography-gutterBottom-118"
              >
                Related reads
              </h2>
              <h2
                className="MuiTypography-root-101 MuiTypography-subheading-108 MuiTypography-gutterBottom-118"
              >
                def
              </h2>
              <div
                style={
                  Object {
                    "display": "flex",
                    "marginBottom": "3%",
                  }
                }
              >
                <div
                  className="MuiAvatar-root-98 MuiAvatar-colorDefault-99"
                  style={
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "green",
                      "height": 50,
                      "width": 50,
                    }
                  }
                >
                  A
                </div>
                <div
                  style={
                    Object {
                      "display": "flex",
                      "flexDirection": "column",
                      "marginLeft": 10,
                    }
                  }
                >
                  <p
                    className="MuiTypography-root-101 MuiTypography-body1-110"
                    style={
                      Object {
                        "marginTop": 6,
                      }
                    }
                  >
                    Author is deleted
                  </p>
                  <p
                    className="MuiTypography-root-101 MuiTypography-body1-110 MuiTypography-colorTextSecondary-124"
                    style={
                      Object {
                        "marginTop": 2,
                      }
                    }
                  >
                    22 September 2018
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-2-59"
  />
</div>
`;

exports[`Article Detail renders Article Detail with isDeletingArticle to be true 1`] = `
<div
  className="MuiGrid-container-1"
  style={
    Object {
      "marginTop": "4%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-2-59"
  />
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-8-65"
  >
    <div
      className="MuiGrid-container-1"
    >
      <div
        style={
          Object {
            "marginBottom": "2%",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "marginBottom": "3%",
            }
          }
        >
          <div
            className="MuiAvatar-root-98 MuiAvatar-colorDefault-99"
            style={
              Object {
                "alignSelf": "center",
                "backgroundColor": "green",
                "height": 50,
                "margin": 10,
                "width": 50,
              }
            }
          >
            A
          </div>
          <div
            style={
              Object {
                "display": "flex",
                "flexDirection": "column",
              }
            }
          >
            <h3
              className="MuiTypography-root-101 MuiTypography-subheading-108"
              style={
                Object {
                  "marginTop": 12,
                }
              }
            >
              Author is deleted
            </h3>
            <p
              className="MuiTypography-root-101 MuiTypography-body1-110 MuiTypography-colorTextSecondary-124"
            >
              22 September 2018
            </p>
          </div>
          <div
            style={
              Object {
                "alignItems": "flex-end",
                "display": "flex",
                "flex": 1,
                "justifyContent": "flex-end",
              }
            }
          >
            <button
              className="MuiButtonBase-root-152 MuiButton-root-126 MuiButton-text-128 MuiButton-outlined-134 MuiButton-sizeSmall-149"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                  "marginRight": 10,
                }
              }
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-127"
              >
                Edit article
              </span>
              <span
                className="MuiTouchRipple-root-155"
              />
            </button>
            <button
              className="MuiButtonBase-root-152 MuiButton-root-126 MuiButton-text-128 MuiButton-outlined-134 MuiButton-sizeSmall-149"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "alignSelf": "center",
                }
              }
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiButton-label-127"
              >
                <div
                  className="MuiCircularProgress-root-204 MuiCircularProgress-colorPrimary-207 MuiCircularProgress-indeterminate-206"
                  role="progressbar"
                  style={
                    Object {
                      "height": 22,
                      "width": 22,
                    }
                  }
                >
                  <svg
                    className="MuiCircularProgress-svg-209"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      className="MuiCircularProgress-circle-210 MuiCircularProgress-circleIndeterminate-212"
                      cx={44}
                      cy={44}
                      fill="none"
                      r={20.2}
                      strokeWidth={3.6}
                      style={Object {}}
                    />
                  </svg>
                </div>
              </span>
              <span
                className="MuiTouchRipple-root-155"
              />
            </button>
          </div>
        </div>
        <h1
          className="MuiTypography-root-101 MuiTypography-display1-105 MuiTypography-gutterBottom-118"
        >
          def
        </h1>
        tester
      </div>
    </div>
    <hr
      className="MuiDivider-root-162"
      style={
        Object {
          "marginBottom": 32,
          "marginTop": 32,
        }
      }
    />
    <div
      className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
    >
      <div
        className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 MuiGrid-grid-md-4-61"
      >
        <div
          id="footer-2"
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
              "flexGrow": 1,
            }
          }
        >
          <div
            className="MuiPaper-root-167 MuiPaper-elevation1-170 MuiPaper-rounded-168 MuiCard-root-166"
            style={
              Object {
                "boxShadow": "0px 0.25px 0.35px 0.35px rgba(0, 0, 0, 0.1)",
                "maxWidth": 385,
              }
            }
          >
            <div
              className="MuiCardMedia-root-194"
              src={undefined}
              style={
                Object {
                  "height": 0,
                  "paddingTop": "56.25%",
                }
              }
            />
            <div
              className="MuiCardContent-root-196"
              style={
                Object {
                  "paddingRight": "2%",
                }
              }
            >
              <h2
                className="MuiTypography-root-101 MuiTypography-caption-111 MuiTypography-gutterBottom-118"
              >
                Related reads
              </h2>
              <h2
                className="MuiTypography-root-101 MuiTypography-subheading-108 MuiTypography-gutterBottom-118"
              >
                def
              </h2>
              <div
                style={
                  Object {
                    "display": "flex",
                    "marginBottom": "3%",
                  }
                }
              >
                <div
                  className="MuiAvatar-root-98 MuiAvatar-colorDefault-99"
                  style={
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "green",
                      "height": 50,
                      "width": 50,
                    }
                  }
                >
                  A
                </div>
                <div
                  style={
                    Object {
                      "display": "flex",
                      "flexDirection": "column",
                      "marginLeft": 10,
                    }
                  }
                >
                  <p
                    className="MuiTypography-root-101 MuiTypography-body1-110"
                    style={
                      Object {
                        "marginTop": 6,
                      }
                    }
                  >
                    Author is deleted
                  </p>
                  <p
                    className="MuiTypography-root-101 MuiTypography-body1-110 MuiTypography-colorTextSecondary-124"
                    style={
                      Object {
                        "marginTop": 2,
                      }
                    }
                  >
                    22 September 2018
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-1-30 MuiGrid-grid-md-2-59"
  />
</div>
`;
